metrics-server/src/metrics.rs

```diff
@@ -0,0 +1,267 @@
+        self.describe(&key, unit, description)
+    }
+
+    fn register_counter(&self, key: &Key, _metadata: &Metadata<'_>) -> Counter {
+        let counter = Arc::new(MetricsCounter::default());
+        self.register(key, Metric::Counter(Arc::clone(&counter)));
+        Counter::from_arc(counter)
+    }
+
+    fn register_gauge(&self, key: &Key, _metadata: &Metadata<'_>) -> Gauge {
+        let gauge = Arc::new(MetricsGauge::default());
+        self.register(key, Metric::Gauge(Arc::clone(&gauge)));
+        Gauge::from_arc(gauge)
+    }
+
+    fn register_histogram(&self, key: &Key, _metadata: &Metadata<'_>) -> Histogram {
+        let hist = Arc::new(MetricsHistogram::default());
+        self.register(key, Metric::Histogram(Arc::clone(&hist)));
+        Histogram::from_arc(hist)
+    }
+}
+
+fn convert_labels_to_prometheus(labels: Vec<Label>) -> Vec<(String, String)> {
+    labels
+        .into_iter()
+        .map(Label::into_parts)
+        .map(|(key, value)| (key.into_owned(), value.into_owned()))
+        .collect()
+}
+
+fn convert_unit_to_prometheus(unit: Unit) -> PrometheusUnit {
+    match unit {
+        Unit::Count => PrometheusUnit::Other("count".to_string()),
+        Unit::Percent => PrometheusUnit::Other("percent".to_string()),
+        Unit::Seconds => PrometheusUnit::Other("seconds".to_string()),
+        Unit::Milliseconds => PrometheusUnit::Other("milliseconds".to_string()),
+        Unit::Microseconds => PrometheusUnit::Other("microseconds".to_string()),
+        Unit::Nanoseconds => PrometheusUnit::Other("nanoseconds".to_string()),
+        Unit::Tebibytes => PrometheusUnit::Other("tebibytes".to_string()),
+        Unit::Gigibytes => PrometheusUnit::Other("gigibytes".to_string()),
+        Unit::Mebibytes => PrometheusUnit::Other("mebibytes".to_string()),
+        Unit::Kibibytes => PrometheusUnit::Other("kibibytes".to_string()),
+        Unit::Bytes => PrometheusUnit::Bytes,
+        Unit::TerabitsPerSecond => PrometheusUnit::Other("terabits per second".to_string()),
+        Unit::GigabitsPerSecond => PrometheusUnit::Other("gigabits per second".to_string()),
+        Unit::MegabitsPerSecond => PrometheusUnit::Other("megabits per second".to_string()),
+        Unit::KilobitsPerSecond => PrometheusUnit::Other("kilobits per second".to_string()),
+        Unit::BitsPerSecond => PrometheusUnit::Other("bits per second".to_string()),
+        Unit::CountPerSecond => PrometheusUnit::Other("count per second".to_string()),
+    }
+}
+
+#[derive(Debug, Default)]
+struct MetricsCounter {
+    inner: prometheus_client::metrics::counter::Counter,
+}
+
+impl CounterFn for MetricsCounter {
+    fn increment(&self, value: u64) {
+        self.inner.inc_by(value);
+    }
+
+    fn absolute(&self, value: u64) {
+        // This technically violates the assumptions of having a monotonically increasing counter.
+        self.inner.inner().store(value, Ordering::SeqCst);
```

Comment:
```suggestion
        self.inner.inner().fetch_max(value, Ordering::Relaxed);
```

This way it's not violating the assumption (and this is also how it's implemented upstream).